sql-plugin/src/main/scala/com/nvidia/spark/rapids/Plugin.scala

```diff
@@ -328,6 +324,34 @@ object RapidsExecutorPlugin {
       zipped.forall { case (e, a) => e <= a }
     }
   }
+
+  private val CUDA_EXCEPTION: String = classOf[ai.rapids.cudf.CudaException].getName
+  private val CUDA_FATAL_EXCEPTION: String = classOf[ai.rapids.cudf.CudaFatalException].getName
+  private val CUDF_EXCEPTION: String = classOf[ai.rapids.cudf.CudfException].getName
+
+  def isCudaException(ef: ExceptionFailure): Boolean = ef.exception match {
+    case Some(ex) =>
+      ex.isInstanceOf[ai.rapids.cudf.CudaException]
+    case None =>
+      ef.description.contains(CUDA_EXCEPTION) ||
+        ef.toErrorString.contains(CUDA_EXCEPTION)
```

Comment:
Is it necessary to scrape the description message now that we're catching it by the class type?  Has the code been tested to verify we are not coming through the message-scraping code path?

Reply:
Hi @jlowe, I encountered a problem: I couldn't find out an easy way to trigger CUDA errors via spark APIs.

Reply:
Hi @jlowe, I tried the above approach, it did creating a fatal CUDA error. And I found that cuDF JNI couldn't indentify (fatal) CUDA errors due to RMM calls. Because CUDA API calls in rmm are wrapped with `RMM_CUDA_TRY_ALLOC` instead of `CUDF_CUDA_TRY`:
```c++
#define RMM_CUDA_TRY_ALLOC(_call)                                                                  \
  do {                                                                                             \
    cudaError_t const error = (_call);                                                             \
    if (cudaSuccess != error) {                                                                    \
      cudaGetLastError();                                                                          \
      auto const msg = std::string{"CUDA error at: "} + __FILE__ + ":" + RMM_STRINGIFY(__LINE__) + \
                       ": " + cudaGetErrorName(error) + " " + cudaGetErrorString(error);           \
      if (cudaErrorMemoryAllocation == error) {                                                    \
        throw rmm::out_of_memory{msg};                                                             \
      } else {                                                                                     \
        throw rmm::bad_alloc{msg};                                                                 \
      }                                                                                            \
    }                                                                                              \
  } while (0)
```

Reply:
> Shall we catch allrmm::bad_alloc as CudaFatalException in cuDF JNI?

rmm::bad_alloc simply means "something else went wrong other than an allocation failure."  It does not necessarily mean a CUDA fatal exception occurred.

However this demonstrates that the CUDA errors are asynchronous, and we're not guaranteed that libcudf is the one processing the error directly (and thus translating it into a CUDA fatal exception).  That means we really have no choice but to fallback to scraping the messages, looking for something that looks like a fatal CUDA error (i.e.: something like Tom's original change).  We can still leverage the cudf exception types, but I don't see a way to avoid scraping the messages in the worst case. 😞

Reply:
Yes, it seems no other way but scraping the error messages, in order to make sure that we catch as many fatal errors as possible. I added back Tom's original approach as a double check over CUDA errors thrown by cuDF/rmm, since libcudf may not be able to catch all fatal errors.

Reply:
Hi @jlowe, I filed a cuDF PR ([10884](https://github.com/rapidsai/cudf/pull/10884)) to do it.